### PR TITLE
roachtest: add support for starting a cluster with store dir snapshot

### PR DESCRIPTION
**Pull request by danhhz**
The motivating examples are things like the 2TB backup test and the
pkg/acceptance allocator tests. I'm still working on creating the 2TB
backup store directories, so this works through a very small example
backup test.

A couple small cleanups while I'm in here: copy `defaultRetryOptions`
from base, shaving a significant time off the compilation of
./pkg/cmd/workload and require "workload" instead of "test" as the
database when loading fixtures.

Release note: None

### CODE REVIEW

**Comment by petermattis** [comment]
Reviewed 4 of 4 files at r1.
Review status: 1 of 14 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/backup.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p3Vm3vwSaBGtp0RIN:-L5p3Vm3vwSaBGtp0RIO:b-j73pu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/backup.go#L45)):*
> ```Go
> 			t.Fatal(err)
> 		}
> 		c.RestoreStoreDirSnapshots(ctx, gen)
> ```

Per my other comment, this feels to abstracted. Perhaps I'm missing the number of such operations we'll need, but I'm wondering why should put this operation in `*cluster` vs looping over the nodes here and explicitly listing the `gs` resources we want to copy to the store directories. 

---

*[pkg/cmd/roachtest/storedir.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p2r3T_MbgMMgaEJml:-L5p2r3T_MbgMMgaEJmm:b-1li3wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/storedir.go#L68)):*
> ```Go
> 			// Assumes cluster configuration of 1 store per node.
> 			storeIdx, numStores := node-1, c.nodes
> 			path := workloadccl.StoreDir(fixtureConfig(), gen, storeIdx, numStores)
> ```

Linking in `workloadccl` here is unfortunate. Seems like there is a lot of abstraction here to figure out a set of `gs` paths. 

---

*[pkg/cmd/roachtest/storedir.go, line 73 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p2b87ygTWBtBBgjbJ:-L5p2b87ygTWBtBBgjbK:bz25udt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/storedir.go#L73)):*
> ```Go
> 			// TODO(dan): Get this path from roachprod instead of hardcoding it.
> 			c.Run(ctx, node, `rm -rf /mnt/data1/cockroach/*`)
> 			c.Run(ctx, node, `mkdir -p /mnt/data1/cockroach`)
> ```

I mentioned on another `roachprod/roachtest` issue that we should add a `{store-dir}` expansion to `roachprod ssh`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850)*

**Comment by danhhz** [comment]
Review status: 1 of 14 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/backup.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p3Vm3vwSaBGtp0RIN:-L5p8BiYckTb-AmidT4-:b-dc0vo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/backup.go#L45)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Per my other comment, this feels to abstracted. Perhaps I'm missing the number of such operations we'll need, but I'm wondering why should put this operation in `*cluster` vs looping over the nodes here and explicitly listing the `gs` resources we want to copy to the store directories. 
</blockquote></details>

I been playing with this for a bit and don't think I've hit on the right abstraction yet, but we need _some_ abstraction. We've more or less been trying the "explict list" strategy and it's been a big big maintenance burden. I had another idea after sending this out that will move this abstraction into workload and make it a better roachtest citizen. Let me take another swing at this

---

*[pkg/cmd/roachtest/storedir.go, line 73 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p2b87ygTWBtBBgjbJ:-L5pAza4lCbaXcfewB1S:br7d5d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/storedir.go#L73)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I mentioned on another `roachprod/roachtest` issue that we should add a `{store-dir}` expansion to `roachprod ssh`. 
</blockquote></details>

gotcha. makes sense

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850)*

**Comment by petermattis** [comment]
Review status: 1 of 14 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/backup.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p3Vm3vwSaBGtp0RIN:-L5pCWPOaHYgWz6tbohg:b-con67e) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/backup.go#L45)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I been playing with this for a bit and don't think I've hit on the right abstraction yet, but we need _some_ abstraction. We've more or less been trying the "explict list" strategy and it's been a big big maintenance burden. I had another idea after sending this out that will move this abstraction into workload and make it a better roachtest citizen. Let me take another swing at this
</blockquote></details>

I'm likely missing context about the maintenance burden. #22859 ported the 2tb restore test to `roachtest` and it was quite clear what was going on. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850)*

**Comment by danhhz** [comment]
Review status: 1 of 14 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/backup.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p3Vm3vwSaBGtp0RIN:-L5pDFpB7OI6nvnYLYg6:b-btwham) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/backup.go#L45)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm likely missing context about the maintenance burden. #22859 ported the 2tb restore test to `roachtest` and it was quite clear what was going on. 
</blockquote></details>

tl;dr I thought about this more and am going to close this PR and try a more incremental approach

The pain stems from three main issues. 1) discoverability. these fixtures are stored in at least 3 different formats, in at least 5 different buckets, on 2 cloud providers. (this is even after nikhil cleaned them up) 2) introspection. it's very hard to say what a given fixture is without spinning up a cluster. there's some info in docs/cloud-resources.md, but it's not comprehensive (for example, how many rows are in the 2tb block_writer dataset and what was the min and max row size used?) 3) regeneration. if none of the existing fixtures are what you need (or there is one but you don't find it because of (1)) then making a new one has traditionally taken at least half a day of hands-on dev time if not more (plus the background time once you get it running)

(3) has largely been solved with roachprod + workload. we need more docs on the pieces but I'm now able to easily make new fixtures. (2) has also been solved for any fixtures made by workload. the path is guaranteed to contain all flags.

I realized what I was doing here was trying to also solve (1), but I wasn't learning from the mistake of pkg/acceptance and was doing it by being prescriptive. I'm going to close this PR and go with the much simpler explicit lists approach and see how far that gets us

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850)*

**Comment by petermattis** [comment]
Review status: 1 of 14 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/backup.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p3Vm3vwSaBGtp0RIN:-L5skfHXYkUu6HqWgmmP:bmyokd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/backup.go#L45)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

tl;dr I thought about this more and am going to close this PR and try a more incremental approach

The pain stems from three main issues. 1) discoverability. these fixtures are stored in at least 3 different formats, in at least 5 different buckets, on 2 cloud providers. (this is even after nikhil cleaned them up) 2) introspection. it's very hard to say what a given fixture is without spinning up a cluster. there's some info in docs/cloud-resources.md, but it's not comprehensive (for example, how many rows are in the 2tb block_writer dataset and what was the min and max row size used?) 3) regeneration. if none of the existing fixtures are what you need (or there is one but you don't find it because of (1)) then making a new one has traditionally taken at least half a day of hands-on dev time if not more (plus the background time once you get it running)

(3) has largely been solved with roachprod + workload. we need more docs on the pieces but I'm now able to easily make new fixtures. (2) has also been solved for any fixtures made by workload. the path is guaranteed to contain all flags.

I realized what I was doing here was trying to also solve (1), but I wasn't learning from the mistake of pkg/acceptance and was doing it by being prescriptive. I'm going to close this PR and go with the much simpler explicit lists approach and see how far that gets us
</blockquote></details>

Ack. Thanks for the details.

To address 2 and 3, can we include some details next to the fixtures about how they were generated and what data they contain?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850)*

**Comment by danhhz** [comment]
Review status: 1 of 14 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/backup.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850#-L5p3Vm3vwSaBGtp0RIN:-L5sxwuVX66BZUgPH531:b1ayzy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e1ec1f0bcfe43d02b2a4859b92ac4cccc2dadb4/pkg/cmd/roachtest/backup.go#L45)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ack. Thanks for the details.

To address 2 and 3, can we include some details next to the fixtures about how they were generated and what data they contain?
</blockquote></details>

Yes, that's a good idea. I'm focused on porting things this week, but I'll add that next time I'm in that code

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22850)*
